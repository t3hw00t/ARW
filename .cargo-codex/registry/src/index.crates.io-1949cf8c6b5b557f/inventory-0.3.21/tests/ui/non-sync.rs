use std::rc::Rc;
use std::thread;

struct Thing(Rc<i32>);

inventory::collect!(Thing);

fn clone_all() {
    for thing in inventory::iter::<Thing> {
        let _ = Rc::clone(&thing.0);
    }
}

fn main() {
    // It would be bad if this were allowed. These threads would race on the
    // nonatomic reference counts.
    let thread1 = thread::spawn(clone_all);
    let thread2 = thread::spawn(clone_all);
    thread1.join().unwrap();
    thread2.join().unwrap();
}
