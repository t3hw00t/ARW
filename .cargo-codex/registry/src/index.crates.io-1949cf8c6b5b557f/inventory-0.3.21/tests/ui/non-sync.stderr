error[E0277]: `Rc<i32>` cannot be shared between threads safely
 --> tests/ui/non-sync.rs:6:21
  |
6 | inventory::collect!(Thing);
  |                     ^^^^^ `Rc<i32>` cannot be shared between threads safely
  |
  = help: within `Thing`, the trait `Sync` is not implemented for `Rc<i32>`
note: required because it appears within the type `Thing`
 --> tests/ui/non-sync.rs:4:8
  |
4 | struct Thing(Rc<i32>);
  |        ^^^^^
note: required by a bound in `Collect`
 --> src/lib.rs
  |
  | pub trait Collect: Sync + Sized + 'static {
  |                    ^^^^ required by this bound in `Collect`
